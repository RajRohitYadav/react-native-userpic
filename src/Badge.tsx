import React, { useEffect, useState } from 'react';
import {
  Animated,
  PixelRatio,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  ViewProps,
  ViewStyle,
} from 'react-native';

import { clamp, getRadius } from './helpers';

const minSize = 15;
const maxSize = 40;

export interface BadgeProps extends ViewProps {
  size?: number;
  color?: string;
  borderRadius?: number | string;
  animate?: boolean;
  value?: number | boolean | string;
  limit?: number | boolean;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
}

export const Badge: React.FC<BadgeProps> = ({
  size = 20,
  color = '#ff3b30',
  borderRadius = '50%',
  animate = true,
  value,
  limit = 99,
  style,
  textStyle,
  ...props
}) => {
  const toValue = value ? 1 : 0;
  const [animatedValue] = useState(() => new Animated.Value(toValue));

  useEffect(() => {
    Animated.spring(animatedValue, {
      tension: 60,
      friction: 6,
      toValue: toValue,
      useNativeDriver: true,
    }).start();
  }, [animatedValue, toValue]);

  if (!value) {
    return null;
  }

  let content = null;
  let height = styles.root.minHeight;

  if (typeof value === 'number' || typeof value === 'string') {
    height = clamp(Math.round(size), minSize, maxSize);

    const fontSize = PixelRatio.roundToNearestPixel(height * 0.6);
    const displayValue = typeof value === 'number' && typeof limit === 'number' && value > limit ? `${limit}+` : value;
    const textStyles = [
      {
        fontSize,
        marginHorizontal: fontSize / 2,
      },
      styles.text,
      textStyle,
    ];

    content = (
      <Text style={textStyles} numberOfLines={1}>
        {displayValue}
      </Text>
    );
  }

  const rootStyles = [
    {
      height,
      minWidth: height,
      backgroundColor: color,
      borderRadius: getRadius(borderRadius, height),
    },
    styles.root,
    style,
  ];

  if (animate) {
    rootStyles.push({ transform: [{ scale: animatedValue as any }] });
  }

  return (
    <Animated.View {...props} style={rootStyles}>
      {content}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  root: {
    minHeight: 10,
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    color: '#fff',
    fontWeight: '400',
    fontFamily: 'System',
    includeFontPadding: false,
    textAlignVertical: 'center',
    backgroundColor: 'transparent',
  },
});
