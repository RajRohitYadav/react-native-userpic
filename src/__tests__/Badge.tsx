import React from 'react';
import { Animated as RNAnimated, PixelRatio, StyleSheet, Text, View } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

const Animated = jest.mocked(RNAnimated);

/**
 * Under test
 */
import Badge, { Props as BadgeProps, BadgePositions } from '../Badge';

const createElement = (props: Partial<BadgeProps>) => <Badge {...props} />;

const createRenderer = (props: Partial<BadgeProps>) => TestRenderer.create(createElement(props));

it('should match snapshot', () => {
  expect(createRenderer({ value: 12 })).toMatchSnapshot();
});

it('should render with a boolean value', () => {
  const { root } = createRenderer({ value: true });
  expect(root.findAllByType(Text).length).toBe(0);
});

it('should render when a value greather than limit', () => {
  const limit = 99;
  const { root } = createRenderer({ value: 100, limit });
  expect(root.findByType(Text).props.children).toBe(`${limit}+`);
});

it('should render customized appearance', () => {
  const props = {
    value: 12,
    size: 24,
    color: 'blue',
    radius: 3,
    style: { minHeight: 5 },
    textStyle: { color: 'red' },
  };
  const { root } = createRenderer(props);
  const wrapper = root.findByType(View);
  const text = root.findByType(Text);

  const fontSize = PixelRatio.roundToNearestPixel(props.size * 0.6);

  expect(wrapper.props.style).toEqual(
    expect.objectContaining({
      height: props.size,
      minWidth: props.size,
      minHeight: props.style.minHeight,
      backgroundColor: props.color,
      borderRadius: props.radius,
    }),
  );

  expect(StyleSheet.flatten(text.props.style)).toEqual(
    expect.objectContaining({
      fontSize,
      color: props.textStyle.color,
      marginHorizontal: fontSize / 2,
    }),
  );
});

Object.values(BadgePositions).map((position) => {
  it(`should render when position = ${position}`, () => {
    const props = { size: 25, value: 12, position, parentSize: 50, parentRadius: 20 };
    const { root } = createRenderer(props);
    const wrapper = root.findByType(View);

    const edgeOffset = props.parentRadius * (1 - Math.sin((45 * Math.PI) / 180));
    const selfOffset = props.size * (0.25 + props.parentRadius / (props.parentSize * 2));
    const offset = PixelRatio.roundToNearestPixel(edgeOffset - selfOffset);
    const [badgeY, badgeX] = position.split('-');

    expect(wrapper.props.style).toEqual(
      expect.objectContaining({
        [badgeY]: offset,
        [badgeX]: offset,
        height: props.size,
        position: 'absolute',
      }),
    );
  });
});

[0, '', false, undefined].map((value) => {
  it(`should not render when value = ${value}`, () => {
    const { root } = createRenderer({ value });
    expect(root.children.length).toBe(0);
  });
});

describe('animation', () => {
  const start = jest.fn();
  const config = {
    toValue: 1,
    friction: 6,
    tension: 50,
    useNativeDriver: true,
  };

  jest.spyOn(Animated, 'spring').mockImplementation(() => ({ start }));

  beforeEach(() => {
    start.mockClear();
  });

  it('should not animate when turned off by the prop', () => {
    const tree = createRenderer({ value: 0, animate: false });

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: 1, animate: false }));
    });

    expect(Animated.spring).not.toBeCalled();
  });

  it('should not animate when numeric value changes', () => {
    const tree = createRenderer({ value: 1 });

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: 2 }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 1 }), config);
    expect(Animated.spring).toBeCalledTimes(1);
    expect(start).toBeCalled();
  });

  it('should animate when zero value changes to number', () => {
    const tree = createRenderer({ value: 0 });

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: 1 }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 0 }), config);
    expect(Animated.spring).toBeCalledTimes(1);
    expect(start).toBeCalled();
  });

  it('should animate when empty value changes to string', () => {
    const tree = createRenderer({ value: '' });

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ value: '!' }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 0 }), config);
    expect(Animated.spring).toBeCalledTimes(1);
    expect(start).toBeCalled();
  });
});
