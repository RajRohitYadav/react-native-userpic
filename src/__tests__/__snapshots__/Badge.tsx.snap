// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ff3b30",
      "borderRadius": 10,
      "flexDirection": "row",
      "height": 20,
      "justifyContent": "center",
      "minWidth": 20,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    ellipsizeMode="clip"
    numberOfLines={1}
    style={
      [
        {
          "backgroundColor": "transparent",
          "color": "#fff",
          "fontFamily": "System",
          "fontSize": 12,
          "fontWeight": "400",
          "includeFontPadding": false,
          "marginHorizontal": 6,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    12
  </Text>
</View>
`;
